Refactor api-versioning.md Developer doc

JeromeJu opened the pull request:
<!-- 🎉🎉🎉 Thank you for the PR!!! 🎉🎉🎉 -->

# Changes
This commit refactors the api-versioning.md developer doc which
previously mixed API versioning and feature versioning. The API-driven
feature versioning was mingled with the process of adding a new CRD
apiVersion.

To have better clarification of the difference between API versioning
and the feature versioning, a new Markdown file feature-versioning is
introduced in light of TEP-0138.

/kind documentation
part of #6965
<!-- Describe your changes here- ideally you can get that description straight from your descriptive commit message(s)! -->

# Submitter Checklist

As the author of this PR, please check off the items in this checklist:

- [x] Has [Docs](https://github.com/tektoncd/community/blob/main/standards.md#docs) if any changes are user facing, including updates to minimum requirements e.g. Kubernetes version bumps
- [n/a] Has [Tests](https://github.com/tektoncd/community/blob/main/standards.md#tests) included if any functionality added or changed
- [x] Follows the [commit message standard](https://github.com/tektoncd/community/blob/main/standards.md#commits)
- [x] Meets the [Tekton contributor standards](https://github.com/tektoncd/community/blob/main/standards.md) (including functionality, content, code)
- [x] Has a kind label. You can add one by adding a comment on this PR that contains `/kind <type>`. Valid types are bug, cleanup, design, documentation, feature, flake, misc, question, tep
- [n/a] Release notes block below has been updated with any user facing changes (API changes, bug fixes, changes requiring upgrade notices or deprecation warnings). See some examples of [good release notes](https://github.com/tektoncd/community/blob/main/standards.md#good-release-notes).
- [n/a] Release notes contains the string "action required" if the change requires additional action from users switching to the new release

# Release Notes

```release-note
NONE
```

Yongxuanzhang commented on docs/developers/feature-versioning.md:
```diff
@@ -0,0 +1,131 @@
+
+  - **Behavioural features**: Features that are not controlled by any specific API fields are considered as behavioural features. They are guarded behind their dedicated behavioural flags. For example, [`results-from`](https://github.com/tektoncd/pipeline/blob/fd4cb4621960715439bf0c0757f5be9cad390568/config/config-feature-flags.yaml#L112C3-L112C38) is a behavioural feature flag.
+
+## API-driven Features
```
are we going to add a section of 'ehavioural features'?

QuanZhang-William replied:
Yeah, as a reader I'm expecting to see a "behavioural feature" section in the doc, and I'm assuming all the content in this section is not applicable to "behavioural features" (but in fact, there are quite some similarity).

QuanZhang-William commented on docs/developers/feature-versioning.md:
```diff
@@ -0,0 +1,131 @@
+# Feature Versioning
+
+There are two types of features in Tekton Pipeline, API-driven features and behavioural features. Both types of features comply to the [feature gates](../../api_compatibility_policy.md#feature-gates) and the [feature graduation process](../../api_compatibility_policy.md#feature-graduation-process) specified in the [API compatibility policy](../../api_compatibility_policy.md).
```
Is there really a point to differentiate the 2 types of feature in the context of per feature flag as eventually all the features will be using individual feature flag (so that you don't really need 2 separate subsetions below)?

I uderstand  that `enable-api-field` will still exist for a while during the migration phase, maybe that is worth mentioning the behavior during & after migration.

JeromeJu replied:
Thanks for the suggestion. Yeah eventually we are going to switch to all dedicated feature flag(NOTE not all will be behind Per-feature flag, for example, some behavioural feature flag would need different values, which is not the same as Per-feature flag i.e. largerResult could take different values instead of true/false) for all new features. But the existing features, which is a combinations of behavioural and API-driven features controlled by `enable-api-fields` will persist for at least ~1 deprecation cycle.

I think it is reasonable to keep the section of behavioural features since they were once the missing pieces in the documentations. And their validations are actually different from Per-feature flags. WDYT?

JeromeJu replied:
TY for the review. I agree with you that users would also like to see a section of behavioural features.

> As a user I would assume all the content in the "API-driven Features" subsection does not apply to "behaviour features", but there are quite some similarities.

Would you mind pointing out what are the similarities. Is it that both of them would use a dedicated feature flag?

QuanZhang-William reviewed (commented):
Thanks @JeromeJu for the update. seems like this is a prerequisite to https://github.com/tektoncd/pipeline/pull/7090?

Did you consider combining the 2 PRs, I feel the change in this PR is not really a complete story of it?

JeromeJu commented on docs/developers/feature-versioning.md:
```diff
@@ -0,0 +1,131 @@
+# Feature Versioning
+
+There are two types of features in Tekton Pipeline, API-driven features and behavioural features. Both types of features comply to the [feature gates](../../api_compatibility_policy.md#feature-gates) and the [feature graduation process](../../api_compatibility_policy.md#feature-graduation-process) specified in the [API compatibility policy](../../api_compatibility_policy.md).
+  - **API-driven features**: Features that are enabled via a specific field in pipeline API. Prior to the switch to per-feature flags, all features gated by `enable-api-fields` are considered as API-driven features. For example, [remote tasks](https://github.com/tektoncd/pipeline/blob/454bfd340d102f16f4f2838cf4487198537e3cfa/docs/taskruns.md#remote-tasks) is an API-driven feature.
```
Restored.

Good point, thanks @QuanZhang-William  this was actually a step-1 for the doc change in https://github.com/tektoncd/pipeline/pull/7090 that I wanted to introduce.

JeromeJu commented on docs/developers/feature-versioning.md:
````diff
@@ -0,0 +1,131 @@
+properly guarded. See the following for an example of a unit test that sets the
+feature gate to test behaviour:
+
+```go
````
Makes sense, updated!

JeromeJu commented:
Thanks for the review @QuanZhang-William ! 
This is a cleanup of docs before introducing the docs in https://github.com/tektoncd/pipeline/pull/7090. There were some updates mingled in this PR which I've removed in the update. TY!

QuanZhang-William reviewed (approved):
Thanks for the PR! Gonna approve it and let more people review!

JeromeJu commented:
Given second thought, I've noticed the confusion it might be for adding the explanation of behavioural flags in this PR. So I removed the part mentioning behavioural flags to avoid being misleading. While this PR mainly serves as a doc cleanup, I have updated the issue tracking this https://github.com/tektoncd/pipeline/issues/6965#issue-1818588964 instead.

Also I've tried to combine this PR in #7090 .

Thanks to the reviews/comments from @QuanZhang-William for letting me understand such potential confusions.

JeromeJu commented:
Closing this as https://github.com/tektoncd/pipeline/pull/7090 has already been merged.
